File: src/tqi/scoring/speed.py
"""Speed sub-score: how fast is transit relative to walking."""

import numpy as np

from tqi.config import TSR_CAR, TSR_WALK
from tqi.scoring.tsr import compute_tsr_matrix, valid_pair_mask

INF = float("inf")


def compute_speed_score(
    distances_km: np.ndarray,
    mean_travel_time: np.ndarray,
    weights: np.ndarray | None = None,
) -> float:
    """Speed score (0-100): normalised mean TSR for reachable pairs.

    If weights are provided (per-origin), uses weighted mean TSR.
    """
    tsr = compute_tsr_matrix(distances_km, mean_travel_time)
    mask = valid_pair_mask(distances_km) & (mean_travel_time < INF)
    if not mask.any():
        return 0.0
    if weights is not None:
        n_origins = tsr.shape[0]
        origin_tsr = np.array([
            float(np.mean(tsr[i, mask[i]])) if mask[i].any() else 0.0
            for i in range(n_origins)
        ])
        mean_tsr = float(np.average(origin_tsr, weights=weights))
    else:
        mean_tsr = float(np.mean(tsr[mask]))
    score = (mean_tsr - TSR_WALK) / (TSR_CAR - TSR_WALK) * 100
    return float(np.clip(score, 0, 100))

File: go/internal/scoring/speed.go
package scoring

import (
	"math"

	"github.com/jasondyck/chwk-tqi/internal/config"
)

// ComputeSpeedScore returns the normalised speed score (0-100).
// Mean TSR for reachable valid pairs is mapped to [0,100] using
// (TSR - TSRWalk) / (TSRCar - TSRWalk) * 100, clamped to [0,100].
func ComputeSpeedScore(distancesKM, meanTravelTime [][]float64) float64 {
	var tsrSum float64
	var count int

	n := len(distancesKM)
	for i := 0; i < n; i++ {
		for j := 0; j < len(distancesKM[i]); j++ {
			if i == j {
				continue
			}
			if !IsValidPair(distancesKM[i][j]) {
				continue
			}
			tsr := ComputeTSR(distancesKM[i][j], meanTravelTime[i][j])
			if tsr <= 0 {
				continue
			}
			tsrSum += tsr
			count++
		}
	}
	if count == 0 {
		return 0
	}
	meanTSR := tsrSum / float64(count)
	score := (meanTSR - config.TSRWalk) / (config.TSRCar - config.TSRWalk) * 100.0
	return math.Max(0, math.Min(100, score))
}
